File: jianzhioffer/firstround/jianzhi_11.py
# class Solution(object): #顺序查找，时间复杂度O(n)
#     def minArray(self, numbers):
#         """
#         :type numbers: List[int]
#         :rtype: int
#         """
#         if len(numbers)==0:
#             return -1
#         elif len(numbers)==1:
#             return numbers[0]
#         else:
#             min = numbers[0]
#             for i in range(1, len(numbers)):
#                 if numbers[i]<min:
#                     min = numbers[i]
#                     break
#             return min

class Solution(object): #二分查找，时间复杂度O(logn)
    def minArray(self, numbers):
        """
        :type numbers: List[int]
        :rtype: int
        """
        if len(numbers)==0:
            return -1
        elif len(numbers)==1:
            return numbers[0]
        else:
            low, high = 0, len(numbers)-1
            while low<high:
                mid = low + (high-low)//2
                if numbers[mid]<numbers[high]:
                    high = mid
                elif numbers[mid]>numbers[high]:
                    low = mid+1
                else:
                    high -= 1
            return numbers[low]


if __name__ == '__main__':
    obj = Solution()
    output = obj.minArray([2,2,2,0,1])
    print(output)

File: jianzhioffer/secondround/jianzhi_11.cpp
#include <vector>
using namespace std;

// //方法一，遍历。时间复杂度O(N)，空间复杂度O(1)
// class Solution {
// public:
//     int minArray(vector<int>& numbers) {
//         if (numbers.empty())
//         {
//             return -1;
//         }
//         int min_num = numbers.back();
//         for (int i = numbers.size()-2; i >= 0; i--)
//         {
//             if (numbers[i]>min_num)
//                 break;
//             else
//                 min_num = numbers[i];
//         }
//         return min_num;
//     }
// };

//方法二，二分查找，时间复杂度O(logN)，空间复杂度O(1)
class Solution {
public:
    int minArray(vector<int>& numbers) {
        if (numbers.empty())
        {
            return -1;
        }
        int low = 0;
        int high = numbers.size()-1;
        while (low<high)
        {
            int pivot = (low+high)/2;
            if (numbers[pivot]<numbers[high])
            {
                high = pivot;
            }
            else if (numbers[pivot]>numbers[high])
            {
                low = pivot + 1;
            }
            else
            {
                high--;
            }        
        }
        return numbers[low];
    }
};

int main(int argc, char const *argv[])
{
    vector<int> n = {3,1};
    Solution obj;
    int result = obj.minArray(n);
    return 0;
}
